File: backend/app/schemas/ai.py
from pydantic import BaseModel
from typing import Optional, Any, Literal, Union
from datetime import datetime
from enum import Enum


class Intent(str, Enum):
    MAGIC_CAPTURE = "magic_capture"
    GENERAL_CHAT = "general_chat"
    LIBRARIAN = "librarian"
    REALITY_CHECK = "reality_check"
    SMART_REVIEW = "smart_review"
    STUCK_BREAKER = "stuck_breaker"
    GOAL_AUDIT = "goal_audit"
    SUNDAY_GARDENER = "sunday_gardener"
    DEEP_PLAN = "deep_plan"


class IntentClassification(BaseModel):
    intent: Intent
    confidence: float
    reasoning: str


class ParsedTask(BaseModel):
    title: str
    description: Optional[str] = None
    priority: Literal["high", "medium", "low"] = "medium"
    due_date: Optional[str] = None
    due_time: Optional[str] = None
    goal_name: Optional[str] = None


class MagicCaptureOutput(BaseModel):
    tasks: list[ParsedTask]
    message: str


class ChatHistoryItem(BaseModel):
    role: str
    content: str


class ChatRequest(BaseModel):
    message: str
    session_id: Optional[str] = None
    client_date: Optional[str] = None
    history: Optional[list[ChatHistoryItem]] = None


class ChatResponse(BaseModel):
    message: str
    session_id: str
    message_metadata: Optional[dict[str, Any]] = None
    actions: Optional[Union[list[dict[str, Any]], dict[str, Any]]] = None
    action_type: Optional[str] = None


class ChatMessageOut(BaseModel):
    id: str
    role: str
    content: str
    message_metadata: dict[str, Any]
    created_at: datetime

    class Config:
        from_attributes = True


class ExecuteTaskData(BaseModel):
    title: str
    description: Optional[str] = None
    priority: Literal["high", "medium", "low"] = "medium"
    time_horizon: Literal["today", "week", "someday"] = "today"
    due_date: Optional[str] = None
    due_time: Optional[str] = None
    goal_id: Optional[str] = None


class ExecuteRequest(BaseModel):
    action_type: str
    data: Union[list[ExecuteTaskData], dict[str, Any]]
    session_id: Optional[str] = None


class ExecuteResponse(BaseModel):
    success: bool
    message: str
    created_ids: Optional[list[str]] = None
    errors: Optional[list[str]] = None

File: frontend/src/app/core/models/ai.ts
import { Priority, TimeHorizon } from './task';

export interface ChatHistoryItem {
  role: string;
  content: string;
}

export interface ChatRequest {
  message: string;
  session_id?: string;
  client_date?: string;
  history?: ChatHistoryItem[];
}

export interface ParsedTask {
  title: string;
  description: string | null;
  priority: Priority;
  time_horizon: TimeHorizon;
  due_date: string | null;
  due_time: string | null;
  goal_id: string | null;
  goal_name: string | null;
}

export interface ChatResponse {
  message: string;
  session_id: string;
  message_metadata?: Record<string, unknown>;
  actions?: ParsedTask[] | SubtaskActions | GoalPlan;
  action_type?: string;
}

export interface ExecuteTaskData {
  title: string;
  description?: string;
  priority: Priority;
  time_horizon: TimeHorizon;
  due_date?: string | null;
  due_time?: string | null;
  goal_id?: string | null;
}

export interface ExecuteRequest {
  action_type: string;
  data: ExecuteTaskData[] | SubtaskActions | GoalPlan;
  session_id?: string;
}

export interface ExecuteResponse {
  success: boolean;
  message: string;
  created_ids?: string[];
  errors?: string[];
}

export interface ChatMessage {
  id: string;
  role: 'user' | 'assistant';
  content: string;
  message_metadata: Record<string, unknown>;
  created_at: string;
}

export interface SubtaskItem {
  title: string;
  description?: string;
}

export interface SubtaskActions {
  parent_task_id: string;
  parent_task_title: string;
  subtasks: SubtaskItem[];
}

export interface PlanTask {
  title: string;
  description: string;
  week_range: string;
  priority: 'high' | 'medium' | 'low';
  estimated_pomodoros: number;
}

export interface GoalPlan {
  goal_title: string;
  goal_description: string;
  goal_color: string;
  duration_weeks: number;
  tasks: PlanTask[];
}

export type ActionData = ParsedTask[] | SubtaskActions | GoalPlan;
